Certificates copy for kubeadm join --control-pane (#713)
kubernetes/enhancements: keps/sig-cluster-lifecycle/20190122-Certificates-copy-for-kubeadm-join--control-plane.md

above and completing the kubeadm solution for creating HA clusters.

### Goals

- Usability: use exactly the same `kubeadm init` and `kubeadm join` commands the users are
  familiar with.
- Security: provide a solution that enables the secure copy of cluster certificates between
  control-plane nodes

## Proposal

### Constraints

The solution described in this proposal is deeply influenced by following constraints:

1. Kubeadm can execute actions only on the machine where it is running e.g. it is not
   possible to execute actions on other nodes.
2. `kubeadm init` and `kubeadm join` are separated actions (executed on separated machines/at
   different times).
3. During the join workflow, kubeadm can access the cluster only using identities with
   limited grants, namely `system:unauthenticated` or `system:node-bootstrapper`.

### Key element of the proposal

At a high level, the proposed solution can be summarized by the following workflow:

**On control plane node 1:**

```bash
kubeadm init --upload-certs
```

The new `--upload-certs` flag will trigger a new kubeadm init phase executing following
actions:

> 1. An encryption key will be generated; such **key will never be stored on cluster**

Reviewer: what is the key length? encryption method?

Reviewer 2: my proposal is 32bytes for key (SHA-256) with AES-256/GCM as the method.
we might want to include the details on that, but others disagree:
https://github.com/kubernetes/enhancements/pull/713#discussion_r249958845

Reviewer 2: also by using a 32byte cryto-secure checksum (e.g. SHA-256) as key to be fed to AES, we can technically use a shorter key on the user side. 
as long as SHA-256 is not a security risk this method can slightly simplify the command line.

Reviewer: > my proposal is 32bytes for key (SHA-256) with AES-256/GCM as the method

+1

> by using a 32byte cryto-secure checksum (e.g. SHA-256) as key to be fed to AES, we can technically use a shorter key on the user side.

if the input to the hash (length and charset) is known, using a hash to expand the data size does not increase entropy, so is not a good idea. to take an extreme example, if the key had a max length of 1 and possible characters 0-9, you could brute force by trying just 10 sha256 hashes.

Reviewer 2: that is true, the extra step will reduce the entropy.

i guess i meant to simplify the command line - e.g.:
`abcdef.1234567890abcdef` (of which we then take the checksum)
vs
`37aa9a83376e16819b4d826732dc0d8a8a310c1b9665f0082ac35e3f04cdd65d`

but we should probably just pass the whole 32bytes.

2. Cluster certificates will be encrypted using the above key

> 3. Encrypted certificates will be stored in a new Kubernetes secret named `kubeadm-certs`;

Reviewer: I assume this includes certificates and their private keys?

Can you enumerate the certificates/keys that will be uploaded:
* apiserver etcd client cert/key?
* apiserver kubelet client cert/key?
* aggregator proxy client cert/key?
* serviceaccount token signing key?
* cluster CA / signing key?
* others?

Does this mean all members joined to an HA cluster share the same client credentials?

   it is important to notice that:

   - This secret is the technical solution that provides a temporary bridge between
     `kubeadm init` and `kubeadm join` commands/between different nodes.

>    - Without the encryption key, this secret contains a harmless blog of bytes.

Reviewer 3: ```suggestion
   - Without the encryption key, this secret contains a harmless bag of bytes.
```
 (or maybe _blob_?)

4. A second bootstrap token will be generated, with a shorter duration than the

>    join token (the TTL-token)

Reviewer 2: should the TTL be an implementation detail outside of the scope of the KEP?

either way, my vote goes for 1h.
non-customizable.

5. The `ownerReference` field in the `kubeadm-certs` secret will be linked to the
   TTL token, thus ensuring automatic deletion of the secret when the TTL-token gets

>    deleted by the `TokenCleaner` controller

Reviewer: this is slightly rube-goldbergesque... is there no stage that finalizes HA control plane setup that could clean this up instead?

Reviewer 4: > this is slightly rube-goldbergesque... is there no stage that finalizes HA control plane setup that could clean this up instead? 

b/c there are may ways an operator can setup HA, we need an auto-cleaning function to prevent potential long term exploits.  We can add clean function to kubeadm, but we still need/want an auto clean.

Reviewer 4: If we had an auto-expiration in object meta, we wouldn't need any of this jiggery, we would auto-expire the secret.

6. RBAC rules will be created to ensure access to the above config map to users
   in the `system:node-bootstrapper` group (the bootstrap tokens).

**On control plane node 2:**

```bash
kubeadm kubeadm join --control-plane --certificate-key={key from step above}
```

The new `--certificate-key` will trigger following actions:

1. The `kubeadm-certs` config will be retrived from the cluster
2. Cluster certificates will be decrypted using the provided key, and
   hen stored on the disk

### Implementation Details

#### The encryption key

> The encryption key will use the same format as Bootstrap Tokens.

Reviewer: I don't think the bootstrap token bit length is sufficient for encrypting the cluster private keys.
~5 bits of entropy per character x 16 characters = ~80 bits of entropy
Something like AES would require 128-256 bits (32-64 characters in hex)

Reviewer 2: +1
it should be a 32byte key so that we can target AES-256.

Author: I removed the bootstrap token part; as noted above, this is linked to the choice of the target encryption method.

See [Authenticating with Bootstrap Tokens](https://kubernetes.io/docs/reference/access-authn-authz/bootstrap-tokens/#token-format)
for more details.

The default lifecycle of the encryption key will be the following:

- a random encryption-key will be created by `kubeadm init --upload-certs`
- the encryption-key will be printed in the output of the `kubeadm init --upload-certs`
  command, but never stored in the cluster
- the encryption-key will expire/become useless as soon as the `kubeadm-certs` will

>   be automatically deleted

Reviewer: > the encryption-key will expire/become useless as soon as the `kubeadm-certs` will be automatically deleted

That's not really accurate... if the content of the kubeadm-certs was already obtained, the encryption-key could later decode it. That's why the bootstrap token length is ok for a credential we can revoke, but not for encryption of data that can be decoded over a longer period of time.

Author: Point taken. I added a note here and also in the risks mitigation chapter where all the potential security concerns are summarized.

3. The user will be allowed to pass the encryption-key to `kubeadm join --control-plane`
   command using the kubeadm config file instead of the `--certificate-key` flag
4. The user will be allowed to generate encryption-key after the first one expires by
   invoking the `kubeadm init phases upload-certs` command (the command will
  re-create/override the existing `kubeadm-certs` secret as well)

#### The kubeadm-certs secret

The `kubeadm-certs` secret will be stored in the `kube-system` namespace; RBAC rules
for ensuring access to the users in the `system:node-bootstrapper` group
(the bootstrap tokens) will be created.

The `kubeadm-certs` secret `ownerReference` attribute will be set to the TTL-token
UID as described in the example below; See [Garbage Collection](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/)
for more details.

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: kubeadm-certs
  namespace: kube-system
  ownerReferences:
  - apiVersion: v1
    controller: true
    blockOwnerDeletion: true
    kind: Secret
    name: bootstrap-token-abcdef
    uid: 84e913dc-191a-11e9-96ae-0242f550e41f
type: Opaque
data:

>   ...

Reviewer: the list of certs/keys and the data map key names will need to be fixed

Author: yes, this list of certs/keys must be well known by the join workflow, but in kubeadm we already have hard-coded constants with certificate names.
I added the list at the end of the paragraphs (the same certs are copied today manually).

```

The `kubeadm-certs` secrets will contain the cluster certificates encrypted
using TBD algorithm / the encryption key described above, and then base64

> encoded as usual.

Reviewer 2: we never spoke about the algorithm.
should this be a technical detail outside of the scope of the KEP?
in my tests i did AES-256 in Galois/Counter Mode (GCM).

Reviewer 5: If we really wanted to we could wrap the algorithm in an interface so we can swap it out if it ends up being insecure (would that even be mathematically possible?). I think at this point it's an implementation detail outside the scope of this KEP.

Reviewer 4: agreed

upload all the necessary certificates (no matter of the cluster variant in use like e.g.
external-CA or external etcd), because the assumption is that, when executed with
the `--upload-certs` flag, kubeadm is delegated by the user to copy all the certificates
required for joining a new control plane node.
In other words, it won't be possible to use kubeadm `--upload-certs` for copying a subset
of certificates only.

#### The TTL-token

The TTL-token is a regular bootstrap token, with a short duration, but without
any assigned usage or groups (see [Authenticating with Bootstrap Tokens](https://kubernetes.io/docs/reference/access-authn-authz/bootstrap-tokens/#bootstrap-token-secret-format)
for more details).

This token is created with the only purpose of triggering automatic deletion of
the `kubeadm-certs` secret when deleted (the two objects will be linked via
`ownerReference` attributes).

The `TokenCleaner` controller will ensure automatic deletion of bootstrap tokens
and therefore also deletion of the TTL-token.

### Risks and Mitigations

In case the cluster administrator uses the `kubeadm init --upload-certs` flag,

> the `kubeadm-certs` secret will store sensitive cluster certificates. Following

Reviewer 3: ```suggestion
the `kubeadm-certs` secret will store sensitive cluster certificates. The following
```

security measures apply to the `kubeadm-certs` secret:

> - the `kubeadm-certs` secret is time lived (with a duration shorter than

Reviewer 3: ```suggestion
- the `kubeadm-certs` secret is only valid for a short period of time (with a duration shorter than
```

  bootstrap tokens)
- if an attacker get access to the cluster when the `kubeadm-certs` secret still
  exists, the cluster certs contained in the secret are protected by RBAC rules
- if an attacker get access to the cluster when the `kubeadm-certs` secret still
  exists with an identity that is authorized to read the `kubeadm-certs` secret,
  the cluster certificates contained in the secret will be protected by the
  encryption algorithm (without the encryption key, the secret contains a harmless

>   blog of bytes)

Reviewer 3: ```suggestion
  bag of bytes)
```
(or maybe _blob_?)

The following security measures apply to the encryption key:

- the encryption key will never be stored on the cluster by kubeadm
- the encryption key will never be transmitted on the wire by kubeadm

> - the encryption key is implicitly time lived (because when the `kubeadm-certs`

Reviewer 3: ```suggestion
- the encryption key is implicitly valid only for a short period of time (because when the `kubeadm-certs`
```

  secret is automatically deleted, the encryption key became useless)
- it will be possible to hide the encryption key from the `kubeadm init` output
  by using `--skip-token-print` or similar flag (but this requires the user
  passing an encryption key to kubeadm using the kubeadm config file), thus
  avoiding the key improperly registered in logs
- it will be possible to use a config file for passing the encryption key to
  `kubeadm join` command, thus avoiding the key improperly registered in
  shell history

Another possible risk derives from kubeadm allowing the user to fully customize
control plane components, including also the possibility to remove the TokenCleaner

> from the list of enabled controllers in the kube-scheduler configuration, and as

Reviewer 3: ```suggestion
from the list of enabled controllers in the kube-scheduler configuration, and, as
```

> a consequence, disable the mechanisms that manages time lived tokens/time lived

Reviewer 3: ```suggestion
a consequence, disable the mechanisms that manages temporarily valid tokens,
```

> `kubeadm-certs` secret/time lived encryption key.

Reviewer 3: ```suggestion
`kubeadm-certs`, and encryption keys.
```

A possible mitigation action for this risk is to enforce the TokenCleaner in case

> when the user opts-in in the upload-certs workflow defined in this document.

Reviewer 2: enforcing might be the right thing to do.
yet i wonder if there is a use case where users **really** don't want to run the TokenCleaner on a cluster. :thinking:

> when the user opts-in in the upload-certs workflow defined in this document.

Reviewer 3: ```suggestion
when the user opts in to the upload-certs workflow defined in this document.
```

Despite all the above measures, there is still a residual risk in case an attacker
manages to steal the encryption key from the cluster administrator and to get access
to the cluster with adequate rights before the `kubeadm-certs` secret is
automatically deleted.

> However, we assume this residual risk could be accepted by cluster administrator

Reviewer 3: ```suggestion
However, we assume this residual risk could be accepted by the cluster administrator
```

in favor of the simplified UX offered by the automatic certificates copy during

> the `kubeadm init` / `kubeadm join --control-plane`.

Reviewer 3: ```suggestion
the `kubeadm init` / `kubeadm join --control-plane` operations.
```

> Nevertheless, as additional mitigation action, the potential risk will be clearly

Reviewer 3: ```suggestion
Nevertheless, as an additional mitigation action, the potential risk will be clearly
```

> stated both in the documentation and in the kubeadm output.

Reviewer 2: should we still provide a mechanic to cancel the serving of certs in such a scenario?

e.g.
`kubeadm init phases upload-certs cancel`

## Design Details

### **Test** Plan

This feature is going to be tested:

> - with unit test

Reviewer 3: ```suggestion
- with unit tests
```

> - with e2e test, if during the v1.14 cycle kubeadm E2E tests are implemented

Reviewer 3: ```suggestion
- with e2e tests, if kubeadm E2E tests are implemented during the v1.14 cycle
```

---
title: Certificates copy for join --control-plane
authors:
  - @fabriziopandini
owning-sig: sig-clusterlifecycle
participating-sigs:
  - 
reviewers:
  - @neolit123 
  - @detiber 
  - @mattmoyer 
  - @chuckha 
  - @liztio 
approvers:
  - @timothysc
  - @luxas
editor: TBD
creation-date: 2019-01-22
last-updated: 2019-01-22

> status: provisional

Reviewer 4: implementable
